### ISSUE TITLE
Borders and spacing broken for JPanel

### ISSUE THREAD

**Issue by Sciss** [mgarin/weblaf#363]
I have to insist that ignoring border settings in a `JPanel` is wrong. This was correctly supported in 1.28 as well as in any other LaF (including Ocean). It seems that furthermore all spacing elements don't work, for example adding `HStrut` or `HGlue` to a `BoxLayout` panel. Here are three screen shots: 1.29-SNAPSHOT (using some dark skin I'm working on but basically it's the default), 1.28 and Ocean:

![screenshot from 2016-03-07 22 17 40](https://cloud.githubusercontent.com/assets/131300/13583758/771eeb0e-e4b3-11e5-8b39-c519b53ea752.png)

![screenshot from 2016-03-07 22 21 17](https://cloud.githubusercontent.com/assets/131300/13583763/7e60e26e-e4b3-11e5-95a2-1d7e2b0fda2d.png)

![screenshot from 2016-03-07 22 21 51](https://cloud.githubusercontent.com/assets/131300/13583764/81e08f20-e4b3-11e5-95a9-8ae204c9435f.png)

**Comment by Sciss** [comment]
I've looked further into the border issue. What happens is that once children are added to the `JPanel`, a `HIERARCHY_CHANGED` event is fired (regular AWT), this goes to a listener in `AbstractDecorationPainter`, then to `AbstractPainer.updateBorder`. It passes a check `isSettingsUpdateAllowed` and then asks

```java
if ( !( old instanceof WebBorder ) || !CompareUtils.equals ( ( ( WebBorder ) old ).getBorderInsets (), border ) )
```

And this will be `true` because I installed a custom `javax.swing.border.TitledBorder`. In my opinion this check must be changed, to something like this:

```java
if ((old == null && border.nonEmpty) || ((old instanceof WebBorder) && !CompareUtils.equals(((WebBorder) old).getBorderInsets(), border)))
```

So a `WebBorder` is installed if currently no border is installed and the required border is non-empty, or if already a `WebBorder` is installed but has different insets

**Comment by Sciss** [comment]
After further checking, the change as proposed above will have problems with some widgets which have bad default borders due to inheritance from BasicPlaf, like textfield. I therefore think we should have a flag for `updateBorder` that indicates whether the call is from `installUI` (first run) or from later calls (e.g. from events). What do you think? I could implement that and test. The other more simple solution is to just handle `WebPanelPainter` differently (here we know that plaf.basic doesn't install a border, so we can have the stronger constraints). So user provided borders would only work for `JPanel`.

**Comment by mgarin** [comment]
This have been fixed now in `styling` branch.

The solution is pretty simple in its nature - any `Border` set which is not `UIResource` will also put (`AbstractPainter` does that) the non-replaceable border flag into the component. That fixes the issue but also allows WebLaF to replace any UI-related borders.

**Comment by mgarin** [comment]
Actually, it was too early to close this issue since `BoxLayout` problems are still there.
I wanted to ask if you can provide any small example representing the issue from the first screenshot?

**Comment by Sciss** [comment]
Yes, I started to make a small self-contained example. It's not finished yet (simply adding HGlue doesn't trigger the problem), and I won't have time probably to do it before tomorrow. I'll post an update here.

**Comment by mgarin** [comment]
Sure, I'll add some other `styling` changes meanwhile.

**Comment by Sciss** [comment]
Here is a minimal example (sorry, it's in Scala but the layout should be straight forward to understand):

```scala
    val topPane = new BoxPanel(Orientation.Horizontal) {
      contents += new Button("Start")
      contents += HGlue
    }

    val bottomPane = new BoxPanel(Orientation.Horizontal) {
      contents += new Label("-------------- Audio Status --------------")
      contents += HGlue
    }

    val boxPane = new BoxPanel(Orientation.Vertical)
    boxPane.contents += topPane
    boxPane.contents += bottomPane

    new MainFrame {
      title     = app.title
      contents  = boxPane
    }
```

So a vertical box panel containing two horizontal box panels with are right-padded by `HGlue`.

This produces with other LaF and with the old WebLaF:

![screenshot from 2016-03-10 14 09 56](https://cloud.githubusercontent.com/assets/131300/13670375/dbed3eb0-e6c9-11e5-8ed3-670f7bdce2ea.png)

Showing the left-alignment of the two horizontal panels.

And with the current snapshot of WebLaF:

![screenshot from 2016-03-10 14 10 15](https://cloud.githubusercontent.com/assets/131300/13670392/f767d18c-e6c9-11e5-81ce-227bef19638c.png)

Showing a (wrong) centred alignment.

**Comment by mgarin** [comment]
The problem was actually in the panel UI which I finally was able to narrow down thanks to your example. I have added a fix for this issue into `styling` branch. 

The problem was in the custom preferred size provided from the UI for all container components which was disabling container layout size calculations. That size is completely unnecessary there due to the fact that it is always layout who provides container size according to its border and children.
